src/esn.inbox.libs/app/services/mailboxes/mailboxes-service.js

```diff
@@ -258,43 +262,55 @@ angular.module('esn.inbox.libs')
     }
 
     function updateSharedMailboxCache() {
+
       return withJmapClient(function(jmapClient) {
-        return jmapClient.mailbox_get({
+        return jmapClient.mailbox_changes({
           accountId: null,
-          ids: null
-        })
-          .then(function(mailboxes) {
-            return mailboxes.list;
-          })
-          .then(function(mailboxList) {
-            return _addSharedMailboxVisibility(_getSharedMailboxes(mailboxList));
-          })
-          .then(function(sharedMailboxList) {
-            const sharedMailboxCache = _getSharedMailboxes(inboxMailboxesCache);
-            const removedSharedFoldersIds = _getDifferenceById(sharedMailboxCache, sharedMailboxList);
-
-            if (!_.isEmpty(removedSharedFoldersIds)) {
-
-              _removeMailboxesFromCache(removedSharedFoldersIds);
-
-              if (removedSharedFoldersIds.includes($state.params.context) === true) {
-                $state.go('unifiedinbox.inbox', { type: '', account: '', context: '' }, { location: 'replace' });
-              }
-            }
-
-            _updateMailboxCache(sharedMailboxList);
-          })
-          .then(function() {
-            $rootScope.$broadcast(INBOX_EVENTS.SHARED_FOLDERS_UPDATED);
+          sinceState: inboxMailboxesCache.state
+        }).then(function(changes) {
+          if (changes.newState !== inboxMailboxesCache.state) {
+            return jmapClient.mailbox_get({
+              accountId: null,
+              ids: null
+            })
+              .then(function(mailboxes) {
+                _updateMailBoxCacheState(mailboxes.state);
+
+                return mailboxes.list;
+              })
+              .then(function(mailboxList) {
+                return _addSharedMailboxVisibility(_getSharedMailboxes(mailboxList));
+              })
+              .then(function(sharedMailboxList) {
+                const sharedMailboxCache = _getSharedMailboxes(inboxMailboxesCache.list);
+                const removedSharedFoldersIds = _getDifferenceById(sharedMailboxCache, sharedMailboxList);
+
+                if (!_.isEmpty(removedSharedFoldersIds)) {
+
+                  _removeMailboxesFromCache(removedSharedFoldersIds);
+
+                  if (removedSharedFoldersIds.includes($state.params.context) === true) {
+                    $state.go('unifiedinbox.inbox', { type: '', account: '', context: '' }, { location: 'replace' });
```

Comment:
How much indentation levels are acceptable?

In my knowledge, method extraction is a powerfull tool not to need 4k screens to read code.

I also allows to better name things.

My take is that maybe those callbacks could be standalone named functions to ease reading...

Reply:
The Indentation base used in the project is space : 2 and it depends on imbrication

Reply:
No that is not what I mean. I count 10 spaces which gives 5 indentation level.

That depicts a big overall complexity of that code. I need to keep 5 branches subdivisions in mind to know what it does.

Generally we loose track from 2-3 indentation levels. 

Extracting things to named separated functions allows to bring the complexity down and enables to keep up with reading.